### PR TITLE
Bug 1942604: Allow-from-router network policy support for ovn-kubernetes

### PR DESCRIPTION
**Pull request by abhat**
This PR cherry-picks commits needed for supporting allow-from-router feature to apply network policy to router endpoints when the endpoint publishing strategy is HostNetwork.

### CODE REVIEW

**Comment by asood-rh** [comment]
@abhat Tested on AWS cluster.

allow-from-router works with old label network.openshift.io/policy-group: ingress but does not work with new label policy-group.network.openshift.io/ingress: "".

allow-from-hostnetwork works fine on the build.

Client Version: 4.7.0
Server Version: 4.6.0-0.ci.test-2021-07-09-190146-ci-ln-wm7tri2-latest
Kubernetes Version: v1.19.2-1083+4c3480dcd4299c-dirty

**Comment by asood-rh** [comment]
All three policies - allow-from-router (old and new label) and allow-from-hostnetwork work fine on a cluster on vSphere (oc get --namespace openshift-ingress-operator ingresscontrollers/default --output jsonpath='{.status.endpointPublishingStrategy.type}'
HostNetwork)

oc version
Client Version: 4.7.0
Server Version: 4.6.0-0.ci.test-2021-07-09-190146-ci-ln-wm7tri2-latest
Kubernetes Version: v1.19.2-1083+4c3480dcd4299c-dirty

File: go-controller/pkg/ovn/gateway_test.go [hunk 1]
```diff
@@ -7,7 +7,9 @@ import (
 	ovntest "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/testing"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
 
+	"github.com/onsi/ginkgo"
```
Comment (martinkennelly) [inline go-controller/pkg/ovn/gateway_test.go:10]: Do you know the reason for duplicate importation of gomega/ginkgo?

Comment (abhat) [reply go-controller/pkg/ovn/gateway_test.go:10]: Probably a result of my local gofmt plugin. In either case dot imports should be only used for testing purpose.

File: go-controller/pkg/ovn/ovn.go [hunk 2]
```diff
@@ -42,10 +42,9 @@ import (
 )
 
 const (
-	egressfirewallCRD                string        = "egressfirewalls.k8s.ovn.org"
-	clusterPortGroupName             string        = "clusterPortGroup"
-	clusterRtrPortGroupName          string        = "clusterRtrPortGroup"
-	egressFirewallDNSDefaultDuration time.Duration = 30 * time.Minute
+	egressfirewallCRD       string = "egressfirewalls.k8s.ovn.org"
+	clusterPortGroupName    string = "clusterPortGroup"
+	clusterRtrPortGroupName string = "clusterRtrPortGroup"
```
Comment (danwinship) [inline go-controller/pkg/ovn/ovn.go:47]: this seems unrelated?

File: go-controller/pkg/util/util.go [hunk 3]
```diff
@@ -40,6 +40,27 @@ const (
...
+	ClusterLBSCTP         = "k8s-cluster-lb-sctp"
+	ClusterLBPrefix       = "k8s-cluster-lb"
+	ClusterIdlingLBPrefix = "k8s-idling-lb"
+	WorkerLBPrefix        = "k8s-worker-lb"
+	WorkerLBTCP           = WorkerLBPrefix + "-tcp"
+	WorkerLBUDP           = WorkerLBPrefix + "-udp"
+	WorkerLBSCTP          = WorkerLBPrefix + "-sctp"
+	GatewayLBTCP          = "TCP_lb_gateway_router"
+	GatewayLBUDP          = "UDP_lb_gateway_router"
+	GatewayLBSCTP         = "SCTP_lb_gateway_router"
```
Comment (danwinship) [inline go-controller/pkg/util/util.go:63]: likewise

File: go-controller/hybrid-overlay/pkg/controller/node_linux.go [hunk 4]
```diff
@@ -631,6 +631,13 @@ func (n *NodeController) syncFlows() {
 	_, _, err = util.ReplaceOFFlows(extBridgeName, flows)
 	if err != nil {
 		klog.Errorf("Failed to add flows, error: %v, flows: %s", err, flows)
+		return
+	}
+
+	// handle resetting ignore learn as we may not have encountered some pods while
+	// iterating through the flows in OVS
+	for _, entry := range n.flowCache {
+		entry.ignoreLearn = false
```
Comment (danwinship) [inline go-controller/hybrid-overlay/pkg/controller/node_linux.go:640]: why is this needed for the allow-from-router stuff?

File: go-controller/pkg/ovn/ovn.go (hunk 2 above)
Comment (abhat) [reply go-controller/pkg/ovn/ovn.go:47]: I had to add those things for getting the utest passing for shared gateway. There were a few unit tests that have been broken on 4.6 branch.

File: go-controller/pkg/util/util.go (hunk 3 above)
Comment (abhat) [reply go-controller/pkg/util/util.go:63]: Same as above. Added for unit-tests.

File: go-controller/hybrid-overlay/pkg/controller/node_linux.go (hunk 4 above)
Comment (abhat) [reply go-controller/hybrid-overlay/pkg/controller/node_linux.go]: It's not. Perhaps a result of incorrect rebasing. Fixed it.

File: go-controller/pkg/ovn/ovn.go (hunk 2 above)
Comment (danwinship) [reply go-controller/pkg/ovn/ovn.go:47]: it seems like it would be better to cherry-pick those fixes separately?

**Comment by danwinship** [comment]
/lgtm
if you decide to split out the unit-test commit before seeking cherry-pick-approved I can re-lgtm
